nuxeo-features/rest-api/nuxeo-rest-api-server/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/OAuth2Object.java

```diff
@@ -58,30 +73,99 @@
 @WebObject(type = "oauth2")
 public class OAuth2Object extends AbstractResource<ResourceTypeImpl> {
 
+    private static final String APPLICATION_JSON_NXENTITY = "application/json+nxentity";
+
+    private static final String TOKEN_DIR = "oauth2Tokens";
+
+    /**
+     * Lists all oauth2 service providers.
+     *
+     * @since 9.2
+     */
+    @GET
+    @Path("provider")
+    public List<NuxeoOAuth2ServiceProvider> getProviders(@Context HttpServletRequest request) throws IOException, RestOperationException {
+        return getProviders();
+    }
+
     /**
      * Retrieves oauth2 data for a given provider.
      */
     @GET
     @Path("provider/{providerId}")
     public Response getProvider(@PathParam("providerId") String providerId,
                                 @Context HttpServletRequest request) throws IOException, RestOperationException {
+        return Response.ok(getProvider(providerId)).build();
+    }
 
-        NuxeoOAuth2ServiceProvider provider = getProvider(providerId);
-
-        Map<String,Object> result = new HashMap<>();
-        result.put("serviceName", provider.getServiceName());
-        result.put("isAvailable", provider.isProviderAvailable());
-        result.put("clientId", provider.getClientId());
-        result.put("authorizationURL", provider.getClientId() == null ? null : provider.getAuthorizationUrl(request));
+    /**
+     * Creates a new OAuth2 service provider.
+     *
+     * @since 9.2
+     */
+    @POST
+    @Path("provider")
+    @Consumes({ APPLICATION_JSON_NXENTITY, "application/json" })
+    public Response addProvider(@Context HttpServletRequest request, NuxeoOAuth2ServiceProvider provider)
+        throws IOException, RestOperationException {
+        if (!((NuxeoPrincipal)getContext().getCoreSession().getPrincipal()).isAdministrator()) {
```

Comment:
You should factor the security check into its own `hasPermission()` method to avoid repeating code. Isn't there an exception you can throw that would be turned into the `FORBIDDEN` by the framework? That would simplify the code slightly (the method would become a simple `checkPermission()` that may throw if not authorized).